File: tests/agent_workloads/adapters/build_test_recovery.mjs
import { cpSync, rmSync } from "node:fs";
import { spawn } from "node:child_process";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

if (process.argv.length !== 3) {
  throw new Error("usage: build_test_recovery INPUT_ROOT");
}

const project = "project";
cpSync(process.argv[2], project, { recursive: true });

function run(program, args) {
  return new Promise((resolve, reject) => {
    const child = spawn(program, args, { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    child.stdout.setEncoding("utf8");
    child.stdout.on("data", (chunk) => {
      stdout += chunk;
    });
    child.stderr.resume();
    child.on("error", reject);
    child.on("close", (code) => resolve({ code, stdout }));
  });
}

try {
  const invalid = await run("node", ["--check", `${project}/invalid.mjs`]);
  if (invalid.code === 0) throw new Error("invalid source passed");

  const checked = await run("node", ["--check", `${project}/valid.mjs`]);
  if (checked.code !== 0) throw new Error("valid source did not build");

  const output = [];
  const originalLog = console.log;
  console.log = (...args) => output.push(args.join(" "));
  try {
    await import(pathToFileURL(resolve(project, "valid.mjs")).href);
  } finally {
    console.log = originalLog;
  }
  if (output.length !== 1) throw new Error("valid source test failed");
  console.log("recovery=ok");
  console.log(`test=${output[0]}`);
} finally {
  rmSync(project, { recursive: true, force: true });
}

File: tests/agent_workloads/adapters/build_test_recovery.py
import shutil
import subprocess
import sys
from pathlib import Path


if len(sys.argv) != 2:
    raise SystemExit("usage: build_test_recovery INPUT_ROOT")

project = Path("project")
shutil.copytree(sys.argv[1], project)
try:
    invalid = subprocess.run(
        ["python3", "-m", "py_compile", str(project / "invalid.py")],
        capture_output=True,
        check=False,
    )
    if invalid.returncode == 0:
        raise SystemExit("invalid source passed")

    checked = subprocess.run(
        ["python3", "-m", "py_compile", str(project / "valid.py")],
        capture_output=True,
        check=False,
    )
    if checked.returncode != 0:
        raise SystemExit("valid source did not build")

    tested = subprocess.run(
        ["python3", str(project / "valid.py")],
        capture_output=True,
        text=True,
        check=False,
    )
    if tested.returncode != 0:
        raise SystemExit("valid source test failed")
    print("recovery=ok")
    print(f"test={tested.stdout.strip()}")
finally:
    shutil.rmtree(project)
